### PR TITLE
Performance of scanning with rpm

### PR DESCRIPTION
**Pull request by yael1994**

### CODE REVIEW

**Comment by yael1994** [comment]
https://trello.com/c/dXuO8zPC/43-improving-the-performance-of-scanning-by-using-file-reads-with-rpm

File: PSSM_score_Peptide/SEQ.cpp [hunk 1]
```diff
@@ -59,5 +59,6 @@ SEQ::SEQ(string & SeqString, const string & SeqName, const double CopyNumber, al
 		_Seq.push_back(alph._alphabetMap[i]);
 	}
 	setSeq_Type();
-	setCopyNmber();
+	//setCopyNmber();
```
Comment (shacharmo) [inline PSSM_score_Peptide/SEQ.cpp]: Why the change is only in this constructor?
What about the other constructor?
Shouldn't this be commented only if used for RPM hits?
Doesn't this affect CalcPSSM_Hits as well? Even if is currently unused in pipeline it should still function correctly

File: PSSM_score_Peptide/main.cpp [hunk 2]
```diff
@@ -361,7 +362,10 @@ void readFileToSeq_array (const string fileName, alphabet& alph, vector<SEQ> &Se
 		{
 			string Seq;
 			string name = currLine.substr(1);
-			
+			if (useCountDuplicate) {
+				auto lastIndex = currLine.find_last_of('_');
+            	count = stod(currLine.substr(lastIndex + 1));
```
Comment (shacharmo) [inline PSSM_score_Peptide/main.cpp:369]: Maybe changing setCopyNumber function would be simpler and supports both cases (if repeats not found, try using RPM)

File: PSSM_score_Peptide/main.cpp [hunk 3]
```diff
@@ -514,11 +518,16 @@ int assignPvalueToPSSMaRRAY(int argc, char *argv[])
...
-		double p_Value = (numberOfRandomPSSM - place +0.0) / numberOfRandomPSSM;
-		if (useFactor & Seq_array.size() > 0) {
-			float factor = float(1000000) / float(Seq_array.size());
+		double p_Value = (numberOfRandomPSSM - place +0.0) / numberOfRandomPSSM;		
+		if (useFactor) {
+			int numberOfSeq = 0;
+			for (size_t k = 0; k < Seq_array.size(); ++k) {
+				numberOfSeq += Seq_array[k]._CopyNumber;
+			}
+			float factor = 1000000 / numberOfSeq;
```
Comment (shacharmo) [inline PSSM_score_Peptide/main.cpp]: No protection if numberOfSeq is 0,
Previous code had the protection

File: PSSM_score_Peptide/main.cpp [hunk 4]
```diff
@@ -514,11 +518,16 @@ int assignPvalueToPSSMaRRAY(int argc, char *argv[])
 		}
 		if (place == -1) place = 0; //so that we get p value = 1 in this case.
 		//cout << "place = " << place << endl;
-		double p_Value = (numberOfRandomPSSM - place +0.0) / numberOfRandomPSSM;
-		if (useFactor & Seq_array.size() > 0) {
-			float factor = float(1000000) / float(Seq_array.size());
+		double p_Value = (numberOfRandomPSSM - place +0.0) / numberOfRandomPSSM;		
+		if (useFactor) {
+			int numberOfSeq = 0;
```
Comment (shacharmo) [inline PSSM_score_Peptide/main.cpp]: This can be expensive and can be "calculated" while reading and returned in readFileToSeq_array

File: PSSM_score_Peptide/main.cpp [hunk 5]
```diff
@@ -514,11 +518,16 @@ int assignPvalueToPSSMaRRAY(int argc, char *argv[])
 		}
 		if (place == -1) place = 0; //so that we get p value = 1 in this case.
 		//cout << "place = " << place << endl;
-		double p_Value = (numberOfRandomPSSM - place +0.0) / numberOfRandomPSSM;
-		if (useFactor & Seq_array.size() > 0) {
-			float factor = float(1000000) / float(Seq_array.size());
+		double p_Value = (numberOfRandomPSSM - place +0.0) / numberOfRandomPSSM;
```
Comment (shacharmo) [inline PSSM_score_Peptide/main.cpp]: why do we need +0.0 (also there's a space missing).
If we want to cast it to double we should do so explicitly

File: hits_cpp/hits.cpp [hunk 6]
```diff
@@ -91,8 +91,12 @@ void memeHits(Meme& meme, AlphabetMap& alphabet, SequencesMap& sequences, int& h
             }
             counter++;
             if (isHit(meme, alphabet, sequencesTypesIter->first, *sequencesIter, verbose)) {
-                meme.addHitSequence(*sequencesIter, isOutputSequences);
-                hits++;
+                if (isUseUniqueRpm){
+                    meme.addHitSequence(*sequencesIter, isOutputSequences, sequncesRpm.find(*sequencesIter)->second);
+                } else {
+                    meme.addHitSequence(*sequencesIter, isOutputSequences);
```
Comment (shacharmo) [inline hits_cpp/hits.cpp]: Should have "1" as third parameter for better readability and symmetry.
If `...->second` is 1 in this case then condition is not required

File: hits_cpp/loadSequences.cpp [hunk 7]
```diff
@@ -34,10 +42,12 @@ SequencesMap loadSequences(string faaPath, int& numSequences, bool verbose) {
             
             getline(file, line);
             sequencesByType->push_back(line);
-            count++;
+            sequncesRpm[line] = unique_rpm;
+            count++ ;
```
Comment (shacharmo) [inline hits_cpp/loadSequences.cpp]: remove space

File: model_fitting/module_wraper.py [hunk 8]
```diff
@@ -305,7 +305,7 @@ def build_classifier(reads_path, motifs_path, model_path, logs_dir, sample2bc, n
 
 def get_faa_file_name_from_path(path, use_mapitope):
     for file_name in os.listdir(path):
-        if file_name.endswith('faa') and 'unique' not in file_name and ('mapitope' in file_name) == use_mapitope:
+        if file_name.endswith('faa') and 'unique' in file_name and ('mapitope' in file_name) == use_mapitope:
```
Comment (shacharmo) [inline model_fitting/module_wraper.py]: Shouldn't we have a flag to support previous code in case RPM implementation has issues?

File: hits_cpp/loadSequences.cpp [hunk 9]
```diff
@@ -16,12 +18,18 @@ SequencesMap loadSequences(string faaPath, int& numSequences, bool verbose) {
...
-            auto lastIndex = line.find_last_of('_');
-            auto seqType = line.substr(lastIndex + 1);
+            replace(line.begin(), line.end(), '_', ' ');  // replace '_' by ' '
+            istringstream iss(line);
+            vector<string> result;
+            for(string s;iss>>s;){
+                result.push_back(s);
+            }
+            auto seqType = result[3];
+            int unique_rpm = stoi(result[7]);
```
Comment (shacharmo) [inline hits_cpp/loadSequences.cpp]: This doesn't support input of non-RPM faa

File: hits_cpp/loadSequences.cpp [hunk 10]
```diff
@@ -16,12 +18,18 @@ SequencesMap loadSequences(string faaPath, int& numSequences, bool verbose) {
     regex pattern("^>.+_(.+)$");
     smatch matches;
     auto end = sequences.end();
-
+    int countAllSequrnces = 0;
     int count = 0;
```
Comment (shacharmo) [inline hits_cpp/loadSequences.cpp]: Why we need both and count and countAllSequences?
Isn't one of them (count) enough?

**Review by shacharmo — CHANGES_REQUESTED** [review]

File: hits_cpp/hits.cpp (hunk 6 above)
Comment (yael1994) [reply hits_cpp/hits.cpp]: ...->second in sequncesRpm does not return 1 when it is shuffles.

File: PSSM_score_Peptide/SEQ.cpp [hunk 11]
```diff
@@ -5,7 +5,6 @@
 
 void SEQ::setCopyNmber()
 {
-	_CopyNumber = 1.0;
```
Comment (shacharmo) [inline PSSM_score_Peptide/SEQ.cpp]: Why remove the line, if input is fasta without repeat or RPM this will fail (number of copies will be 0)

File: hits_cpp/hits.cpp [hunk 12]
```diff
@@ -272,6 +272,7 @@ int main(int argc, char *argv[])
         ("shufflesPercent", "Percent from shuffle with greatest number of hits (0-1)", cxxopts::value<float>()->default_value("0.2"))
         ("shufflesDigits", "Number of digits after the point to print in scanning files", cxxopts::value<int>()->default_value("2"))
         ("useFactor", "To multiply by factor hits for normalization", cxxopts::value<bool>()->default_value("false"))
+        ("useUniqueFaaScanning", "Performance of scanning script with rpm faa file", cxxopts::value<bool>()->default_value("false"))
```
Comment (shacharmo) [inline hits_cpp/hits.cpp]: Name should include RPM, unique is not clear enough.
p-val include the name `useRpmFaaScanning` which is clearer

File: model_fitting/module_wraper.py [hunk 13]
```diff
@@ -44,6 +44,7 @@
     "type_machines_to_stop" : "type_machines_to_stop",
     "name_machines_to_stop" : "name_machines_to_stop",
     "no_rpm_factor": "no_rpm_factor",
+    "use_unique_faa_scanning": "use_unique_faa_scanning",
```
Comment (shacharmo) [inline model_fitting/module_wraper.py]: unique => rpm

File: model_fitting/module_wraper.py [hunk 14]
```diff
@@ -303,9 +304,10 @@ def build_classifier(reads_path, motifs_path, model_path, logs_dir, sample2bc, n
         f.write(' '.join(argv) + '\n')
 
 
-def get_faa_file_name_from_path(path, use_mapitope):
+def get_faa_file_name_from_path(path, use_mapitope, use_unique_faa_scanning):
     for file_name in os.listdir(path):
-        if file_name.endswith('faa') and 'unique' in file_name and ('mapitope' in file_name) == use_mapitope:
+        if file_name.endswith('faa') and (('unique' in file_name and use_unique_faa_scanning) or ('unique' not in file_name and not use_unique_faa_scanning)) \
```
Comment (shacharmo) [inline model_fitting/module_wraper.py]: ```('unique' in file_name) == use_unique_faa_scaning``` is shorter and clearer

**Review by shacharmo — CHANGES_REQUESTED** [review]

File: model_fitting/module_wraper.py (hunk 14 above)
Comment (yael1994) [reply model_fitting/module_wraper.py]: use_rpm_faa_scanning means to use the file unique_rpm.faa, but the `'unique' in file_name` find the right file between all the faa files in the folder.

File: PSSM_score_Peptide/SEQ.cpp (hunk 11 above)
Comment (yael1994) [reply PSSM_score_Peptide/SEQ.cpp]: We set the _CopyNumber to 1 in the constructor or to number from RPM (it is parameter of the constructor ):

`SEQ::SEQ(string & SeqString, const string & SeqName, const double CopyNumber, alphabet& alph) :  _Seq_Name(SeqName), _CopyNumber(CopyNumber), _alph(alph) {
	for (size_t k = 0; k < SeqString.length(); ++k) {
		char i = SeqString.at(k);
		_Seq.push_back(alph._alphabetMap[i]);
	}
	setSeq_Type();
	setCopyNmber();
};
`
if we keep the line it overrides the _CopyNumber. I test the code with the RPM file and regular faa file and it is works good.

Comment (shacharmo) [reply PSSM_score_Peptide/SEQ.cpp]: Maybe we should call setCopyNumber only when needed and not always

File: model_fitting/module_wraper.py (hunk 14 above)
Comment (shacharmo) [reply model_fitting/module_wraper.py]: OK but ` (('unique' in file_name and use_unique_faa_scanning) or ('unique' not in file_name and not use_unique_faa_scanning))` still equals to `('unique' in file_name) == use_unique_faa_scaning`

File: PSSM_score_Peptide/SEQ.cpp (hunk 11 above)
Comment (yael1994) [reply PSSM_score_Peptide/SEQ.cpp]: We use `void SEQ::setCopyNmber()` in two options -  1. calculate pval scanning. 2. calculate shuffle scanning - > but we don't use this code anymore because we use hits_pcc instead. 

when I run the calculate pval scanning it is not going inside the `if (regex_search(_Seq_Name, match, rgx))` 
so I don't see why we need this function. we set the number of copies in the constructor.

Comment (shacharmo) [reply PSSM_score_Peptide/SEQ.cpp]: If we don't need the function at all then remove it BUT it is necessary for other operations such as cutoffs, we don't want to break existing interfaces and find out the damage later.

To resolve, add a boolean parameter of `isSetCopyNumber` to constructor with default of true.
only if `isSetCopyNumber` is true then call `setCopyNumber` function.
When not required set the parameter to false

**Review by shacharmo — APPROVED** [review]
